src/com/ansorgit/plugins/bash/lang/parser/misc/RedirectionParsing.java

```diff
@@ -143,7 +143,9 @@ private boolean parseSingleRedirect(BashPsiBuilder builder, boolean inCheckMode,
 
             ParserUtil.markTokenAndAdvance(builder, HEREDOC_START_ELEMENT);
 
-            builder.getParsingState().pushHeredocMarker();
+            if (!builder.getParsingState().getHeredocMarkersIndexSet().contains(builder.rawTokenIndex())) {
```

Comment:
Does is make sense to move the logic of `only add if not yet added` into `pushHeredocMarker`? That would allow to remove `getHeredocMarkersIndexSet` and to simplify this part.